linux/runtime.go

```diff
@@ -0,0 +1,197 @@
+package linux
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/docker/containerd"
+	"github.com/docker/containerd/api/services/shim"
+	"github.com/docker/containerd/api/types/container"
+	"github.com/docker/containerd/api/types/mount"
+	"github.com/docker/containerd/log"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	runtimeName    = "linux"
+	configFilename = "config.json"
+)
+
+func init() {
+	containerd.RegisterRuntime(runtimeName, New)
+}
+
+func New(root string) (containerd.Runtime, error) {
+	if err := os.MkdirAll(root, 0700); err != nil {
+		return nil, err
+	}
+	c, cancel := context.WithCancel(context.Background())
+	return &Runtime{
+		root:          root,
+		events:        make(chan *containerd.Event, 2048),
+		eventsContext: c,
+		eventsCancel:  cancel,
+	}, nil
+}
+
+type Runtime struct {
+	root string
+
+	events        chan *containerd.Event
+	eventsContext context.Context
+	eventsCancel  func()
+}
+
+func (r *Runtime) Create(ctx context.Context, id string, opts containerd.CreateOpts) (containerd.Container, error) {
+	path, err := r.newBundle(id, opts.Spec)
+	if err != nil {
+		os.RemoveAll(path)
+		return nil, err
+	}
+	s, err := newShim(path)
+	if err != nil {
+		os.RemoveAll(path)
+		return nil, err
+	}
+	if err := r.handleEvents(s); err != nil {
+		os.RemoveAll(path)
+		return nil, err
+	}
+	sopts := &shim.CreateRequest{
+		ID:       id,
+		Bundle:   path,
+		Runtime:  "runc",
+		Stdin:    opts.IO.Stdin,
+		Stdout:   opts.IO.Stdout,
+		Stderr:   opts.IO.Stderr,
+		Terminal: opts.IO.Terminal,
+	}
+	for _, m := range opts.Rootfs {
+		sopts.Rootfs = append(sopts.Rootfs, &mount.Mount{
+			Type:    m.Type,
+			Source:  m.Source,
+			Options: m.Options,
+		})
+	}
+	if _, err := s.Create(ctx, sopts); err != nil {
+		os.RemoveAll(path)
+		return nil, err
+	}
+	return &Container{
+		id:   id,
+		shim: s,
+	}, nil
+}
+
+func (r *Runtime) Delete(ctx context.Context, c containerd.Container) error {
+	lc, ok := c.(*Container)
+	if !ok {
+		return fmt.Errorf("container cannot be cast as *linux.Container")
+	}
+	if _, err := lc.shim.Delete(ctx, &shim.DeleteRequest{}); err != nil {
+		return err
+	}
+	lc.shim.Exit(ctx, &shim.ExitRequest{})
+	return r.deleteBundle(lc.id)
+}
+
+func (r *Runtime) Containers() ([]containerd.Container, error) {
+	dir, err := ioutil.ReadDir(r.root)
+	if err != nil {
+		return nil, err
+	}
+	var o []containerd.Container
+	for _, fi := range dir {
+		if !fi.IsDir() {
+			continue
+		}
+		c, err := r.loadContainer(fi.Name())
```

Comment:
This going to connect to the shim(s) every time it's called.

Wouldn't it be better to keep a map of the existing containers within the runtime instance struct since this could technically be called several time?

**edit**: ok, I see that the supervisor is the only one calling it. Would eventually would need a doc on the interface that this routine could "start from scratch" all the time